src/Lean/Elab/Tactic/BVDecide/Frontend/BVDecide.lean

```diff
@@ -0,0 +1,167 @@
+  for (bitVecVar, bitMap) in sparseMap.toArray do
+    let mut value : Nat := 0
+    let mut currentBit := 0
+    for (bitIdx, bitValue) in bitMap.toList do
+      assert! bitIdx == currentBit
+      if bitValue then
+        value := value ||| (1 <<< currentBit)
+      currentBit := currentBit + 1
+    let atomExpr := atomsAssignment.get! bitVecVar
+    finalMap := finalMap.push (atomExpr, ⟨BitVec.ofNat currentBit value⟩)
+  return finalMap
+
+structure UnsatProver.Result where
+  proof : Expr
+  lratCert : LratCert
+
+abbrev UnsatProver := BVLogicalExpr → Std.HashMap Nat Expr → MetaM UnsatProver.Result
+
+def lratBitblaster (cfg : TacticContext) (bv : BVLogicalExpr)
+    (atomsAssignment : Std.HashMap Nat Expr) :
+    MetaM UnsatProver.Result := do
+  let entry ←
+    withTraceNode `bv (fun _ => return "Bitblasting BVLogicalExpr to AIG") do
+      -- lazyPure to prevent compiler lifting
+      IO.lazyPure (fun _ => bv.bitblast)
+  let aigSize := entry.aig.decls.size
+  trace[bv] s!"AIG has {aigSize} nodes."
+
+  if cfg.graphviz then
+    IO.FS.writeFile ("." / "aig.gv") <| AIG.toGraphviz entry
+
+  let (cnf, map) ←
+    withTraceNode `sat (fun _ => return "Converting AIG to CNF") do
+      -- lazyPure to prevent compiler lifting
+      IO.lazyPure (fun _ =>
+        let (entry, map) := entry.relabelNat'
+        let cnf := AIG.toCNF entry
+        (cnf, map)
+      )
+
+  let res ←
+    withTraceNode `sat (fun _ => return "Obtaining external proof certificate") do
+      runExternal cnf cfg.solver cfg.lratPath cfg.trimProofs cfg.timeout cfg.binaryProofs
+
+  match res with
+  | .ok cert =>
+    let proof ← cert.toReflectionProof cfg bv ``verifyBVExpr ``unsat_of_verifyBVExpr_eq_true
+    return ⟨proof, cert⟩
+  | .error assignment =>
+    let reconstructed := reconstructCounterExample map assignment aigSize atomsAssignment
+    let mut error := m!"The prover found a potential counter example, consider the following assignment:\n"
+    for (var, value) in reconstructed do
+      error := error ++ m!"{var} = {value.bv}\n"
+    throwError error
+
+def reflectBV (g : MVarId) : M (BVLogicalExpr × (Expr → M Expr)) := g.withContext do
+  let hyps ← getLocalHyps
+  let sats ← hyps.filterMapM SatAtBVLogical.of
+  if sats.size = 0 then
+    let mut error := "None of the hypotheses are in the supported BitVec fragment.\n"
+    error := error ++ "There are two potential fixes for this:\n"
+    error := error ++ "1. If you are using custom BitVec constructs simplify them to built-in ones.\n"
+    error := error ++ "2. If your problem is using only built-in ones it might currently be out of reach.\n"
+    error := error ++ "   Consider expressing it in terms of different operations that are better supported."
+    throwError error
+  let sat := sats.foldl (init := SatAtBVLogical.trivial) SatAtBVLogical.and
+  return (sat.bvExpr, sat.proveFalse)
+
+def _root_.Lean.MVarId.closeWithBVReflection (g : MVarId) (unsatProver : UnsatProver) :
```

Comment:
I'm not sure I like the dot notation here, this will pollute autocomplete for all tactic writers.